### DISCUSSION TITLE
Need help in adding Active Directory (AD) authentication through Curl.

### DISCUSSION THREAD

**Discussion by GuruDev-2021** [curl/curl#9876] [Q&A] [answered]
Version of Curl used: 7.78
Environment: Windows 11 / Windows 10
Build environment: Visual Studio

I am using Curl in a client server project where client uses Curl to do HTTP/HTTPS based network calls. All these days I was using Digest Authentication. 

Since the server-side implementation supports Active Directory(AD) authentication through KRBv5, I want to explore from the Client application side also to support AD authentication through Curl.

Note: Our client application uses OpenWSman library (https://github.com/Openwsman) for which in-turn uses curl. 
Sample code from WS Man:

`

	void * init_curl_transport()
	{
		CURL *curl;
		curl = curl_easy_init();
		curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, wsman_transport_get_verify_peer(cl) ? 1 : 0);
		curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, wsman_transport_get_verify_host(cl) ? 2 : 0);
		curl_easy_setopt(curl, CURLOPT_PROXY, cl->proxy_data.proxy);
		curl_easy_setopt(curl, CURLOPT_TIMEOUT, cl->transport_timeout);
		curl_easy_setopt(curl, CURLOPT_PROXYUSERPWD, cl->proxy_data.proxy_auth);
		curl_easy_setopt(curl, CURLOPT_CRLFILE, cl->authentication.crl_file);
		curl_easy_setopt(curl, CURLOPT_CAPATH, cl->authentication.capath);
		curl_easy_setopt(curl, CURLOPT_CAINFO, cl->authentication.cainfo);
		curl_easy_setopt(curl, CURLOPT_SSL_CTX_DATA, (void *)cl->authentication.certificatethumbprint);
		curl_easy_setopt(curl, CURLOPT_SSLKEY, cl->authentication.sslkey);
		curl_easy_setopt(curl, CURLOPT_SSLCERT, cl->authentication.sslcert);
		curl_easy_setopt(curl, CURLOPT_SSLVERSION, CURL_SSLVERSION_DEFAULT);
	}`

`

	void wsmc_handler()
	{
		curl = init_curl_transport();
		curl_easy_setopt(curl, CURLOPT_USE_SSL, CURLUSESSL_ALL);
		curl_easy_setopt(curl, CURLOPT_PIPEWAIT, 1L);
		curl_easy_setopt(curl, CURLOPT_URL, cl->data.endpoint);
		curl_easy_setopt(curl, CURLOPT_PRIVATE, cl->data.endpoint);
		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_handler);
		curl_easy_setopt(curl, CURLOPT_WRITEDATA, response);
		char content_type[64] = { 0 };
		snprintf(content_type, 64, "Content-Type: application/soap+xml;charset=%s", UTF-8);
		headers = curl_slist_append(headers, content_type);
		snprintf(usag, usag_len, "User-Agent: %s", Openwsman);
		headers = curl_slist_append(headers, usag);
		curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
		r = curl_easy_setopt(curl, CURLOPT_POSTFIELDS, buf);
		curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, len);
		while (1)
		{
			_user = wsmc_get_user(cl);
			_pass = wsmc_get_password(cl);
			curl_easy_setopt(curl, CURLOPT_HEADEROPT, CURLHEADER_SEPARATE);
			if (_user && _pass && cl->data.auth_set)
			{
				r = curl_easy_setopt(curl, CURLOPT_HTTPAUTH, cl->data.auth_set);
				upwd = u_strdup_printf("%s:%s", _user, _pass);
				r = curl_easy_setopt(curl, CURLOPT_USERPWD, upwd);
			}
			curl_easy_setopt(curl, CURLOPT_VERBOSE, 1);
			r = curl_easy_perform(curl);
			if (r != CURLE_OK)
			{
				cl->fault_string = u_strdup(curl_easy_strerror(r));
				curl_err("curl_easy_perform failed");
				goto DONE;
			}
			r = curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
			if (r != CURLE_OK)
			{
				cl->fault_string = u_strdup(curl_easy_strerror(r));
				curl_err("curl_easy_getinfo(CURLINFO_RESPONSE_CODE) failed");
				goto DONE;
			}
			switch (http_code)
			{
				case 401:
					// The server requires authentication.
					break;
			}
			if (iDone == 1)
			{
				break;
			}
			r = curl_easy_getinfo(curl, CURLINFO_HTTPAUTH_AVAIL, &auth_avail);
			if (r != CURLE_OK)
			{
				cl->fault_string = u_strdup(curl_easy_strerror(r));
				curl_err("curl_easy_getinfo(CURLINFO_HTTPAUTH_AVAIL) failed");
				goto DONE;
			}
			cl->data.auth_set = reauthenticate(cl, cl->data.auth_set, auth_avail,
				&cl->data.user, &cl->data.pwd);
		u_buf_append(con->response, u_buf_ptr(response), u_buf_len(response));
	DONE:
		curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
		cl->response_code = http_code;
		cl->last_error = convert_to_last_error(r);
		curl_slist_free_all(headers);
		return;
	}`

Questions:
1) Do you have sample implementations of AD with curl? Which OpenSource project has best compatibility for Active Directory(AD)?
2) Has any other open-source project used KRB5 with Curl, that you know of?

**Comment by ganeshkamath89** [comment, marked as answer]
Hi @GuruDev-2021,
I am using OpenWSMan with Kerberos authentiation from Curl on Windows. I am linking my Curl project with libCurl.

There are two options:
1) Building libCurl with USE_WINDOWS_SSPI preprocessor.
Here KRB5 is not used as Windows SSPI is a self contained Kerberos provider.

2) Building libCurl with KRB5 library requires macro HAVING_GSSAPI preprocessor.
I faced issue with the latest KRB5 1.20.1, so I am using the older KRB5 1.18.1 released last year.
Your issue might be related to this particular version of KRB5.

My recommendation is to go with USE_WINDOWS_SSPI preprocessor as OpenWSMan already calls the right APIs to perform Kerberos authentication from client side via Curl.

My Curl version is 7.86.

Hope this helps.

**Reply by GuruDev-2021** [reply]
Hi Ganesh,

Thanks for the information, I could see curl with OpenWSMan working with the steps mentioned above. Curl with Active Directory is started working after building with **USE_WINDOWS_SSPI**. It would be helpful if these things can be captured some where in Curl building document or readme file.

This discussion thread can be closed.

Thanks,
Guru
